src/v/security/authorizer.h

```diff
@@ -290,21 +292,111 @@ class authorizer final {
               bool(_allow_empty_matches));
         }
 
-        // check for deny
-        if (auto entry = acls.find(
-              operation, principal, host, acl_permission::deny);
-            entry.has_value()) {
-            return auth_result::acl_match(
-              principal, host, operation, resource_name, false, *entry);
+        auto check_access =
+          [this, &acls, &operation, &host, &resource_name](
+            const security::acl_principal& user,
+            std::optional<const security::acl_principal_base*> role,
+            acl_permission perm) -> std::optional<auth_result> {
+            assert(
+              !role
+              || *role != nullptr && (*role)->type() == principal_type::role);
```

Comment:
> Can you elaborate a bit? Seems like strictly a logic error to me (vs something that should never happen but _could_). 

Our vassert runs in release and does nice logging and backtrace.

Reply:
Yeah I know, I was trying to avoid paying for it in release builds (since I think it suggests logic error). But if I have to sell it, it's probably not worth worrying about. Happy to change.

Reply:
done